File: arrays/arrays-from-scratch.js
// Build an array from scratch using classes in javascript
class MyArray {
  constructor() {
    this.length = 0;
    this.data = {};
  }

  get(index) {
    return this.data[index];
  }

  push(item) {
    this.data[this.length] = item;
    this.length++;
    return item;
  }

  pop() {
    const lastItem = this.data[this.length - 1];
    delete this.data[this.length - 1];
    this.length--;
    return lastItem;
  }

  delete(index) {
    const deletedItem = this.data[index];
    delete this.data[index];
    this.shiftItems(index);
    delete this.data[this.length - 1];
    this.length--;
    return deletedItem;
  }

  shiftItems(index) {
    for (let i = index; i < this.length - 1; i++) {
      this.data[i] = this.data[i + 1];
    }
  }
}

const newArray = new MyArray();
console.log(newArray.get(0));
console.log(newArray.push("Hi!"));
console.log(newArray.push("Cokesman"));
console.log(newArray.push("MB"));
console.log(newArray.push("BJ"));
console.log(newArray.push("You"));
console.log(newArray.push("are"));
console.log(newArray.push("awesome"));
// console.log(newArray.pop());
// console.log(newArray.pop());
console.log(newArray.delete(2));
console.log(newArray);

File: arrays/python-implementation/arrays-from-scratch.py
class MyArray:
    def __init__(self):
        self.length = 0
        self.data = {}

    def __str__(self):
        return str(self.__dict__)

    def get(self, index):
        return self.data[index]

    def pop(self):
        last_item = self.data[self.length - 1]
        del self.data[self.length - 1]
        self.length -= 1
        return last_item

    def push(self, item):
        self.length += 1
        self.data[self.length - 1] = item
        return self.data 

    def delete(self, index):
        deleted_item = self.data[index]
        del self.data[index]
        self.length -= 1
        self.shift_items(index)
        del self.data[self.length]
        return (deleted_item, self.data)

    def shift_items(self, index):
        for i in range(index, self.length):
            self.data[i] = self.data[i + 1]

    def insert(self, index, item):
        for i in range(index+1, self.length+1):
            self.data[i] = self.data[i - 1]

        self.data[index] = item 
        self.length += 1 
        return self.data




newArray = MyArray()
# print(newArray.get(0))
print(newArray.push("Hi!"))
print(newArray.push("Cokesman"))
print(newArray.push("MB"))
print(newArray.push("BJ"))
print(newArray.push("You"))
print(newArray.push("are"))
print(newArray.push("awesome"))
print(newArray.pop())
print(newArray.pop())
print(newArray.delete(2))
print(newArray.insert(3, "Ishow"))



